app/assets/javascripts/components/utils.js.coffee

```diff
@@ -34,10 +34,62 @@ class @Utils
     body?(modal.querySelector('.modal-body'))
     $(modal).modal('show')
 
-  @handleDryRunButton: (button, data = $(button.form).serialize()) ->
+  @handleDryRunButton: (button, data = if button.form then $(':input[name!="_method"]', button.form).serialize() else '') ->
     $(button).prop('disabled', true)
+    cleanup = -> $(button).prop('disabled', false)
+
+    url = $(button).data('action-url')
+    with_event = $(button).data('with-event')
+
+    if with_event is 'no'
+      return @invokeDryRun(url, data, cleanup)
+
+    Utils.showDynamicModal """
+      <h5>Event to send#{if with_event is 'maybe' then ' (Optional)' else ''}</h5>
+      <form class="dry-run-form" method="post">
+        <div class="form-group">
+          <textarea rows="10" name="event" class="payload-editor" data-height="200">
+            {}
+          </textarea>
+        </div>
+        <div class="form-group">
+          <input value="Dry Run" class="btn btn-primary" type="submit" />
+        </div>
+      </form>
+      """,
+      body: (body) =>
+        form = $(body).find('.dry-run-form')
+        window.setupJsonEditor(form.find('.payload-editor'))
+        form.submit (e) =>
+          e.preventDefault()
+          json = $(e.target).find('.payload-editor').val()
+          json = '{}' if json == ''
+          try
+            payload = JSON.parse(json)
+            throw true unless payload.constructor is Object
+            if Object.keys(payload).length == 0
+              json = ''
+            else
+              json = JSON.stringify(payload)
+          catch
+            alert 'Invalid JSON object.'
+            return
+          if json == ''
+            if with_event is 'yes'
+              alert 'Event is required for this agent to run.'
+              return
+            dry_run_data = data
+          else
+            dry_run_data = "event=#{encodeURIComponent(json)}&#{data}"
+          $(body).closest('[role=dialog]').on 'hidden.bs.modal', =>
```

Comment:
My understanding is that you're doing this to detect when the modal is closed, right?  Could we use the `onHide` callback of `showDynamicModal` instead?